### ISSUE TITLE
Remove restriction on '/' prefixes

### ISSUE THREAD

**Issue by bmeck** [nodejs/modules#540]
Given that we have self referential packages, and `"imports"` in `package.json`. In addition since `//` is reserved I think we should remove the reservation since it will by the default URL parser be treated as protocol relative.

**Comment by jkrems** [comment]
To be clear, this is about direct import specifiers (https://nodejs.org/api/esm.html#esm_import_specifiers), e.g.:

```js
// imports file:///usr/local/etc/mytool.config.mjs, assuming import.meta.url has file:// protocol.
import config from '/usr/local/etc/mytool.config.mjs';
```

I think that makes sense. Given `"imports"`, I don't see a strong reason to reserve design space here (IIRC package-relative was the only thing ever discussed for this spot).

**Comment by ljharb** [comment]
Wait, why would we want to allow this? What are the use cases for root-absolute specifiers?

I can't conceive of a time in CJS where requiring from a /-prefixed path wasn't a bug (or bad design, as was later revealed when it was ran on a different machine).

**Comment by bmeck** [comment]
@ljharb because it adds consistency and I cannot think of a reason to disallow these specifiers on concrete concerns if we are not reserving them for future features. Any future feature using those prefixes not matching the generally accepted behavior of URL resolution would be bad so it isn't like we can use the feature for other purposes beyond pinning to some kind of root.

**Comment by ljharb** [comment]
I absolutely agree that *if* `/` is allowed as a non-`//` prefix, it would have to function as an absolute path, but I don't think "consistency" is a good argument to enable something that is almost always a footgun.

**Comment by bmeck** [comment]
@ljharb why is it a footgun? is there clear reasoning here or is it just a linting issue?

**Comment by ljharb** [comment]
It's a footgun because it's not project-portable or machine-portable; any time you have code that uses `/`, it's highly likely to break on someone else's machine (including a different docker image, host vs guest, if you migrate from one computer to a new one, etc).

**Comment by bmeck** [comment]
@ljharb that statement is true for any given one off script that isn't installed via a package manager and uses bare specifiers for things. I do see utility in various things like the ability to `import '/etc/config/foo.json';` regardless.

**Comment by jkrems** [comment]
I have definitely imported things using absolute paths. If it's top-level application code that will only ever run in a docker container, I don't see anything wrong with importing something (like a config file) from an absolute path. I haven't seen strong evidence that this behavior caused issues in `require` but have seen people use it there for - what I would consider - valid reasons. But that's admittedly anecdotal.

**Comment by DerekNonGeneric** [comment]
I have some thoughts on this topic, hopefully it will help in coming to a conclusion. :)

---

**Behavior &num; 1**
*LAMP server*

Assuming content-type headers are set correctly on a standard-issue LAMP web server, the following should be possible.

```js
import * as ns from '/foo.mjs';
```

This would import `foo.mjs` located at `/var/www/html/foo.mjs` a.k.a. `https://example.com/foo.mjs`.

AFAIK (note: I don't use node as a web server) Node.js doesn't have a directory like this, so we can't really mimic this behavior.

---

**Behavior &num; 2**:
*Flags*

IMO, Node.js *should* be able to access the filesystem root. After all, one of the major advantages Node.js has over Deno is that Node.js *does not* put access to the [filesystem, network, or environment behind flags like Deno](https://deno.land/manual/getting_started/permissions#permissions-list). Because of this lack of flag necessity, dependencies can access the filesystem root. (BTW, I think this lack of flag + package ecosystem is what allows Node.js to compete w/ Python for back-end scripting language market share.)

---

Funny enough, this (access to the filesystem root) is already working as I would expect on OSs w/ POSIX paths, so changing it to something different now might be a bit awkward.

In closing, this is just how it seems to me at the moment, there could be things I'm unaware of (i.e., impossible to determine filesystem root dir in some rare OSs, et al.) &mdash; please take w/ grain of salt there could be other things I'm unaware of.

**Comment by jkrems** [comment]
> AFAIK (note: I don't use node as a web server) Node.js doesn't have a directory like this, so we can't really mimic this behavior.

Node.js does have a "directory" like this - because modules in node use the `file:` protocol and host-relative resolution of `/` works the same here as it does in HTTP URLs. Resolving `/zapp` relative to `file:///foo/bar` is well-defined across platforms. Most importantly it's *not* generally a path relative to the system root, not even on OSs with POSIX paths. Resolving `/a%20b.js` will not resolve to the file with POSIX path `/a%20b.js`. It will resolve to a "completely" different path: `/a b.js`. Those are two different files.

The fact that Apache _may_ serve some of the `http:` URLs from some directory on disk is a special case with one particular configuration. E.g. Apache could serve some URLs from a constant string in the config or from running a CGI script or from various directories depending on URL patterns. In terms of resolution, that shouldn't really matter.

**Comment by ljharb** [comment]
That node's implementation uses URLs doesn't mean that node module specifiers are *conceptually* URLs - they're file paths (which happen to be a subset of URLs).

**Comment by bmeck** [comment]
@ljharb the files are URLs, not paths, they do have fragments taken into consideration.

**Comment by ljharb** [comment]
I realize they are URLs. I'm saying most users will never think of them that way.

**Comment by bmeck** [comment]
@ljharb people are using this preservation of URL aspects for things like cache busting or ferrying meta-data already, it isn't something we can decouple.

**Comment by jkrems** [comment]
> That node's implementation uses URLs doesn't mean that node module specifiers are conceptually URLs - they're file paths (which happen to be a subset of URLs).

I think it would be dangerous if we allow users to think that they are file paths. If allowing `/` reinforces that notion, then I'd consider it an argument to keep `/` "forbidden". Code like `import(path.resolve(filename)))` may _look_ at first glance like it should work but it will not be portable or safe. It will only work if both `filename` and the working directory happen to only contain URL-safe characters that are interpreted the same way in both filenames and `file:` URLs. True for many common filenames but not true for arbitrary filenames.

**Comment by DerekNonGeneric** [comment]
> I think it would be dangerous if we allow users to think that they are file paths. If allowing / reinforces that notion, then I'd consider it an argument to keep / "forbidden".

Something perhaps worth noting is that [MDN's very first example of a module specifier](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/import#Import_an_entire_modules_contents) in their documentation page on static import statements uses a leading `/`. 😕 Could it be that this is just a path in this case?

For some reason it seems to me that we may have a feature missing that's supposed to control what a specifier with a leading `/` is supposed to be relative to (e.g. somehow being able to set a [DocumentRoot](https://httpd.apache.org/docs/current/mod/core.html#documentroot) of sorts)? It would function by passing that directory to something akin to [`require.resolve()`'s `options.paths <string[]>`](https://nodejs.org/dist/latest-v14.x/docs/api/modules.html#modules_require_resolve_request_options) to determine where to start looking. 🤔

**Comment by ljharb** [comment]
@DerekNonGeneric that's because MDN is geared towards the web, not node, and on the web, that's a domain-relative URL.

**Comment by jkrems** [comment]
> I feel very strongly that if a leading / is allowed, it can only ever mean "the filesystem root", because that's what it already means in CJS.

Yes, for `file:` referrers, it would roughly be "the file system root". But if we explain it as "just like CommonJS", we're asking for trouble. Because while it may relative to the file system root, it's not a file system path (!). The following will load two different files, even though the specifiers are interpreted relative to the file system root:

```js
require('/tmp/a%20b.js'); // the file "/tmp/a%20.js"
import('/tmp/a%20b.js'); // the file "/tmp/a b.js"
```

If we don't think people can differentiate between those two different paths, maybe we shouldn't allow a leading slash in `import`.

**Comment by DerekNonGeneric** [comment]
@ljharb, I'm not opposed to keeping the leading `/` to mean relative to the filesystem root as you like. In fact, I quite like it that way too. I was simply proposing that there may be value in being able to override that default behavior via config (perhaps a pjson field).

Essentially it would be opt-in to override the default. I think the Node.js web server community could see some real value in this, but I'm still researching the status-quoe atm. :)

/cc @wesleytodd as Express users may be into this?

**Comment by bmeck** [comment]
@DerekNonGeneric that seems like you just want to run a `chroot` around node, not that node would actually need to do anything.

**Comment by DerekNonGeneric** [comment]
> @DerekNonGeneric that seems like you just want to run a chroot around node, not that node would actually need to do anything.

@bmeck, that's not cross-platform, I don't understand how that is the suggestion here.

**Comment by ljharb** [comment]
Wait, what does `/` in CJS on Windows map to now??

**Comment by DerekNonGeneric** [comment]
FS Root, just like you said, right?

**Comment by DerekNonGeneric** [comment]
```sh
C:\>node --eval "require('/index.js')"
C:\index.js:3
export {};
^^^^^^

SyntaxError: Unexpected token 'export'
```

**Comment by ljharb** [comment]
Right, but "c:" isn't necessarily the fs root on windows - where linux has `/`, and all volumes are mounted under that, windows has up to 26 (i think) volume roots, and i'm not aware there *is* any pathable location that contains all of them.

Perhaps it maps to the root of the working directory's drive?

**Comment by bmeck** [comment]
@DerekNonGeneric 

> that's not cross-platform, I don't understand how that is the suggestion here.

It works on everywhere but windows natively and windows has workarounds for this via cygwin, windows subsystem for linux, etc. The point was that the request doesn't seem to be something that node needs to specifically handle, and the details of chroot are quite complex.

**Comment by DerekNonGeneric** [comment]
Node is very close to working very well on Windows to my knowledge. Of course, there is the problem of libuv being implemented wrong and messing up colors in cmd.exe and PowerShell. It's close to being something I can actually ship cross-platform production software on. Treating Windows as a red-headed step-child is a mistake and support should be considered equally important IMO.

**Comment by bmeck** [comment]
@DerekNonGeneric there is no point that I stated that windows is second class.

**Comment by jkrems** [comment]
After reading the other recent thread, I think my bigger concern is with \*nix platforms where people may confuse `/some/file/url/path` with `/some/os/file/name` and then run into issues with the different encoding. It seemed very hard to get the point across that there are (some) practical differences between the two because they look so similar.

I would be in favor of a clear warning in our resolver when resolving a host-relative URL against a `file:` URL, suggesting people use an explicit protocol. Not a really strong opinion (there's plenty of common file naming patterns where it's fine) but it may be worth it.
